File: iterations/tabletest/iteration-45/eval-27-convert-from-testng/outputs/src/test/java/com/example/ShippingCostCalculatorTest.java
package com.example;

import org.junit.jupiter.api.DisplayName;
import org.tabletest.junit.Description;
import org.tabletest.junit.TableTest;
import org.tabletest.junit.TypeConverter;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ShippingCostCalculatorTest {

    private final ShippingCostCalculator calculator = new ShippingCostCalculator();

    @DisplayName("Calculates the base rate from zone and weight tier")
    @Description("""
        Dimensions fixed at 1x1x1 cm so the billable weight always equals the actual
        weight (billable weight is covered separately, see
        chargesByDimensionalWeightWhenItExceedsActual). No fragile, insurance, or
        hazmat options. Carrier fixed at DHL — price does not depend on carrier
        (see chargesSameRegardlessOfCarrier).
        """)
    @TableTest("""
        Scenario                            | Zone         | Weight (kg) | Base Rate?
        EU standard, up to 1 kg              | EU/standard  | {0.5, 1}    | 5.00
        EU standard, over 1 kg up to 5 kg    | EU/standard  | {1.01, 5}   | 7.50
        EU standard, over 5 kg up to 15 kg   | EU/standard  | {5.01, 15}  | 12.50
        EU standard, over 15 kg              | EU/standard  | {15.01, 25} | 20.00
        EU express, up to 1 kg               | EU/express   | {0.5, 1}    | 8.00
        EU express, over 1 kg up to 5 kg     | EU/express   | {1.01, 5}   | 12.00
        EU express, over 5 kg up to 15 kg    | EU/express   | {5.01, 15}  | 20.00
        EU express, over 15 kg               | EU/express   | {15.01, 25} | 32.00
        US standard, up to 1 kg              | US/standard  | {0.5, 1}    | 7.00
        US standard, over 1 kg up to 5 kg    | US/standard  | {1.01, 5}   | 10.50
        US standard, over 5 kg up to 15 kg   | US/standard  | {5.01, 15}  | 17.50
        US standard, over 15 kg              | US/standard  | {15.01, 25} | 28.00
        US express, up to 1 kg               | US/express   | {0.5, 1}    | 11.00
        US express, over 1 kg up to 5 kg     | US/express   | {1.01, 5}   | 16.50
        US express, over 5 kg up to 15 kg    | US/express   | {5.01, 15}  | 30.00
        US express, over 15 kg               | US/express   | {15.01, 25} | 45.00
        """)
    void calculatesBaseRateByZoneAndWeight(ShippingZone zone, double weight, BigDecimal baseRate) {
        BigDecimal result = calculator.calculateShippingCost(
                zone, weight, List.of(1, 1, 1), new PackageOptions(), Carrier.DHL);
        assertEquals(0, result.compareTo(baseRate));
    }

    @DisplayName("Charges by dimensional weight when it exceeds the actual weight")
    @Description("""
        Zone fixed at EU standard, no other options. Billable weight is the greater
        of the actual weight and the dimensional weight (length x width x height / 5000).
        """)
    @TableTest("""
        Scenario                                     | Weight (kg) | Dimensions   | Total Cost?
        Actual weight exceeds the dimensional weight  | 12          | [10, 10, 10] | 12.50
        Dimensional weight exceeds the actual weight  | 1           | [40, 30, 5]  | 7.50
        """)
    void chargesByDimensionalWeightWhenItExceedsActual(double weight, List<Integer> dimensions, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(
                zone, weight, dimensions, new PackageOptions(), Carrier.DHL);
        assertEquals(0, result.compareTo(totalCost));
    }

    @DisplayName("Adds an oversize fee when a dimension exceeds the threshold")
    @Description("Zone fixed at EU standard, 3 kg (base rate $7.50), no other options.")
    @TableTest("""
        Scenario                            | Dimensions   | Total Cost?
        At the oversize threshold (100 cm)  | [100, 5, 5]  | 7.50
        Just over the threshold (101 cm)    | [101, 5, 5]  | 17.50
        """)
    void addsOversizeFeeForDimensionsOverThreshold(List<Integer> dimensions, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, dimensions, new PackageOptions(), Carrier.DHL);
        assertEquals(0, result.compareTo(totalCost));
    }

    @DisplayName("Applies a 15% multiplier for fragile packages")
    @Description("Zone fixed at EU standard, 3 kg, dimensions 10x10x10 cm (base rate $7.50).")
    @TableTest("""
        Scenario     | Options         | Total Cost?
        Not fragile  | [:]             | 7.50
        Fragile      | [fragile: true] | 8.625
        """)
    void appliesFragileMultiplier(PackageOptions options, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, List.of(10, 10, 10), options, Carrier.DHL);
        assertEquals(0, result.compareTo(totalCost));
    }

    @DisplayName("Charges the greater of the computed premium or the insurance minimum")
    @Description("""
        Zone fixed at EU standard, 3 kg, dimensions 10x10x10 cm (base rate $7.50).
        Insurance premium is 0.6% of the insured value, with a $3.00 minimum.
        """)
    @TableTest("""
        Scenario                    | Options              | Total Cost?
        No insurance declared       | [:]                  | 7.50
        Premium at the minimum      | [insuredValue: 500]  | 10.50
        Premium above the minimum   | [insuredValue: 600]  | 11.10
        """)
    void addsInsuranceSurchargeAboveMinimum(PackageOptions options, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, List.of(10, 10, 10), options, Carrier.DHL);
        assertEquals(0, result.compareTo(totalCost));
    }

    @DisplayName("Adds a flat fee for hazmat handling, ignoring other handling instructions")
    @Description("Zone fixed at EU standard, 3 kg, dimensions 10x10x10 cm (base rate $7.50).")
    @TableTest("""
        Scenario                          | Options               | Total Cost?
        No handling instruction           | [:]                   | 7.50
        Non-hazmat handling instruction   | [handling: priority]  | 7.50
        Hazmat handling instruction       | [handling: hazmat]    | 15.50
        """)
    void addsHazmatHandlingFee(PackageOptions options, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, List.of(10, 10, 10), options, Carrier.DHL);
        assertEquals(0, result.compareTo(totalCost));
    }

    @DisplayName("Adds flat fees before the fragile multiplier, then adds insurance last")
    @Description("""
        Zone fixed at EU standard, 3 kg, dimensions 10x10x10 cm (base rate $7.50). Shows
        an interaction neither single-rule table above shows: flat fees (oversize,
        hazmat) are added to the base before the fragile multiplier is applied, while
        insurance is added afterward, on top of the multiplied total.
        """)
    @TableTest("""
        Scenario                                           | Options                              | Total Cost?
        Hazmat fee is added before the fragile multiplier  | [handling: hazmat, fragile: true]    | 17.825
        Insurance is added after the fragile multiplier    | [fragile: true, insuredValue: 200]   | 11.625
        """)
    void combinesSurchargesInFixedOrder(PackageOptions options, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, List.of(10, 10, 10), options, Carrier.DHL);
        assertEquals(0, result.compareTo(totalCost));
    }

    @DisplayName("Charges the same price regardless of carrier")
    @Description("Zone fixed at EU express, 3 kg, dimensions 30x20x15 cm (base rate $12.00), no other options.")
    @TableTest("""
        Scenario                       | Carrier            | Total Cost?
        Same price for every carrier   | {DHL, UPS, FEDEX}  | 12.00
        """)
    void chargesSameRegardlessOfCarrier(Carrier carrier, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "express");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, List.of(30, 20, 15), new PackageOptions(), carrier);
        assertEquals(0, result.compareTo(totalCost));
    }

    @TypeConverter
    public static ShippingZone parseZone(String value) {
        String[] parts = value.split("/");
        return new ShippingZone(parts[0], parts[1]);
    }

    @TypeConverter
    public static PackageOptions parsePackageOptions(Map<String, String> config) {
        PackageOptions options = new PackageOptions();
        if (Boolean.parseBoolean(config.getOrDefault("fragile", "false"))) {
            options.setFragile(true);
        }
        if (config.containsKey("insuredValue")) {
            options.setInsuredValue(new BigDecimal(config.get("insuredValue")));
        }
        if (config.containsKey("handling")) {
            options.setHandling(config.get("handling"));
        }
        return options;
    }
}

File: iterations/tabletest/iteration-89/eval-25-convert-from-spock/outputs/src/test/kotlin/com/example/ShippingCostCalculatorTest.kt
package com.example

import org.junit.jupiter.api.Assertions.assertEquals
import org.tabletest.junit.Description
import org.tabletest.junit.TableTest
import org.tabletest.junit.TypeConverter
import java.math.BigDecimal

class ShippingCostCalculatorTest {

    private val calculator = ShippingCostCalculator()

    @TableTest(
        """
        Scenario                                        | Zone        | Weight (kg) | Dimensions (cm) | Options | Carrier | Cost?
        EU standard, weight at the 1kg boundary         | EU standard | 1.0         | [10, 10, 10]    | [:]     | DHL     | 5.00
        EU standard, weight just past the 1kg boundary  | EU standard | 1.01        | [10, 10, 10]    | [:]     | DHL     | 7.50
        EU standard, weight at the 5kg boundary         | EU standard | 5.0         | [10, 10, 10]    | [:]     | DHL     | 7.50
        EU standard, weight just past the 5kg boundary  | EU standard | 5.01        | [10, 10, 10]    | [:]     | DHL     | 12.50
        EU standard, weight at the 15kg boundary        | EU standard | 15.0        | [10, 10, 10]    | [:]     | DHL     | 12.50
        EU standard, weight just past the 15kg boundary | EU standard | 15.01       | [10, 10, 10]    | [:]     | DHL     | 20.00
        EU express, light package                       | EU express  | 0.5         | [10, 10, 10]    | [:]     | DHL     | 8.00
        EU express, medium package                      | EU express  | 3.0         | [10, 10, 10]    | [:]     | DHL     | 12.00
        EU express, heavy package                       | EU express  | 10.0        | [10, 10, 10]    | [:]     | DHL     | 20.00
        EU express, very heavy package                  | EU express  | 25.0        | [10, 10, 10]    | [:]     | DHL     | 32.00
        US standard, light package                      | US standard | 0.5         | [10, 10, 10]    | [:]     | DHL     | 7.00
        US standard, medium package                     | US standard | 3.0         | [10, 10, 10]    | [:]     | DHL     | 10.50
        US standard, heavy package                      | US standard | 10.0        | [10, 10, 10]    | [:]     | DHL     | 17.50
        US standard, very heavy package                 | US standard | 25.0        | [10, 10, 10]    | [:]     | DHL     | 28.00
        US express, light package                       | US express  | 0.5         | [10, 10, 10]    | [:]     | DHL     | 11.00
        US express, medium package                      | US express  | 3.0         | [10, 10, 10]    | [:]     | DHL     | 16.50
        US express, heavy package                       | US express  | 10.0        | [10, 10, 10]    | [:]     | DHL     | 30.00
        US express, very heavy package                  | US express  | 25.0        | [10, 10, 10]    | [:]     | DHL     | 45.00
        """
    )
    fun `computes the base rate from zone and weight`(
        zone: ShippingZone, weightKg: Double, dimensions: List<Int>,
        options: PackageOptions, carrier: Carrier, cost: BigDecimal
    ) {
        assertCost(cost, calculator.calculateShippingCost(zone, weightKg, dimensions, options, carrier))
    }

    @TableTest(
        """
        Scenario                                     | Zone        | Weight (kg) | Dimensions (cm) | Options | Carrier | Cost?
        Actual weight exceeds the dimensional weight | EU standard | 3.0         | [10, 10, 10]    | [:]     | DHL     | 7.50
        Dimensional weight exceeds the actual weight | EU standard | 1.0         | [70, 50, 10]    | [:]     | DHL     | 12.50
        """
    )
    fun `charges for the larger of actual and dimensional weight`(
        zone: ShippingZone, weightKg: Double, dimensions: List<Int>,
        options: PackageOptions, carrier: Carrier, cost: BigDecimal
    ) {
        assertCost(cost, calculator.calculateShippingCost(zone, weightKg, dimensions, options, carrier))
    }

    @TableTest(
        """
        Scenario                                | Zone        | Weight (kg) | Dimensions (cm) | Options | Carrier | Cost?
        Dimensions at the oversize threshold    | EU standard | 3.0         | [100, 5, 5]     | [:]     | DHL     | 7.50
        Length just past the oversize threshold | EU standard | 3.0         | [101, 5, 5]     | [:]     | DHL     | 17.50
        Width just past the oversize threshold  | EU standard | 3.0         | [5, 101, 5]     | [:]     | DHL     | 17.50
        """
    )
    fun `adds a flat fee for oversize packages`(
        zone: ShippingZone, weightKg: Double, dimensions: List<Int>,
        options: PackageOptions, carrier: Carrier, cost: BigDecimal
    ) {
        assertCost(cost, calculator.calculateShippingCost(zone, weightKg, dimensions, options, carrier))
    }

    @TableTest(
        """
        Scenario                 | Zone        | Weight (kg) | Dimensions (cm) | Options         | Carrier | Cost?
        Package handled normally | EU standard | 3.0         | [10, 10, 10]    | [:]             | DHL     | 7.50
        Package marked fragile   | EU standard | 3.0         | [10, 10, 10]    | [fragile: true] | DHL     | 8.625
        """
    )
    fun `applies a multiplier for fragile packages`(
        zone: ShippingZone, weightKg: Double, dimensions: List<Int>,
        options: PackageOptions, carrier: Carrier, cost: BigDecimal
    ) {
        assertCost(cost, calculator.calculateShippingCost(zone, weightKg, dimensions, options, carrier))
    }

    @TableTest(
        """
        Scenario                            | Zone        | Weight (kg) | Dimensions (cm) | Options             | Carrier | Cost?
        No insured value declared           | EU standard | 3.0         | [10, 10, 10]    | [:]                 | DHL     | 7.50
        Insurance premium below the minimum | EU standard | 3.0         | [10, 10, 10]    | [insuredValue: 499] | DHL     | 10.50
        Insurance premium above the minimum | EU standard | 3.0         | [10, 10, 10]    | [insuredValue: 501] | DHL     | 10.506
        """
    )
    fun `adds an insurance premium for insured packages`(
        zone: ShippingZone, weightKg: Double, dimensions: List<Int>,
        options: PackageOptions, carrier: Carrier, cost: BigDecimal
    ) {
        assertCost(cost, calculator.calculateShippingCost(zone, weightKg, dimensions, options, carrier))
    }

    @TableTest(
        """
        Scenario                         | Zone        | Weight (kg) | Dimensions (cm) | Options              | Carrier | Cost?
        No handling instructions         | EU standard | 3.0         | [10, 10, 10]    | [:]                  | DHL     | 7.50
        Non-hazmat handling instructions | EU standard | 3.0         | [10, 10, 10]    | [handling: standard] | DHL     | 7.50
        Hazmat handling instructions     | EU standard | 3.0         | [10, 10, 10]    | [handling: hazmat]   | DHL     | 15.50
        """
    )
    fun `adds a flat fee for hazmat handling`(
        zone: ShippingZone, weightKg: Double, dimensions: List<Int>,
        options: PackageOptions, carrier: Carrier, cost: BigDecimal
    ) {
        assertCost(cost, calculator.calculateShippingCost(zone, weightKg, dimensions, options, carrier))
    }

    @TableTest(
        """
        Scenario                        | Zone       | Weight (kg) | Dimensions (cm) | Options | Carrier           | Cost?
        Any carrier gives the same cost | EU express | 3.0         | [10, 10, 10]    | [:]     | {DHL, UPS, FEDEX} | 12.00
        """
    )
    fun `charges the same rate regardless of carrier`(
        zone: ShippingZone, weightKg: Double, dimensions: List<Int>,
        options: PackageOptions, carrier: Carrier, cost: BigDecimal
    ) {
        assertCost(cost, calculator.calculateShippingCost(zone, weightKg, dimensions, options, carrier))
    }

    @Description(
        """
        Neither single-concern table above shows how the surcharges combine. These rows pin the
        order: flat fees (oversize, hazmat) are added first, the fragile multiplier is applied to
        that subtotal, and the insurance premium is added last, after the multiplier.
        """
    )
    @TableTest(
        """
        Scenario                                                    | Zone        | Weight (kg) | Dimensions (cm) | Options                            | Carrier | Cost?
        Flat fees are added before the fragile multiplier           | EU standard | 3.0         | [101, 5, 5]     | [fragile: true, handling: hazmat]  | DHL     | 29.325
        The fragile multiplier applies before the insurance premium | EU standard | 3.0         | [10, 10, 10]    | [fragile: true, insuredValue: 200] | DHL     | 11.625
        """
    )
    fun `applies flat fees and multipliers in a fixed order`(
        zone: ShippingZone, weightKg: Double, dimensions: List<Int>,
        options: PackageOptions, carrier: Carrier, cost: BigDecimal
    ) {
        assertCost(cost, calculator.calculateShippingCost(zone, weightKg, dimensions, options, carrier))
    }

    private fun assertCost(expected: BigDecimal, actual: BigDecimal) {
        assertEquals(0, expected.compareTo(actual), "expected $expected but was $actual")
    }
}

@TypeConverter
fun parseZone(value: String): ShippingZone {
    val (region, speed) = value.split(" ")
    return ShippingZone(region, speed)
}

@TypeConverter
fun parseOptions(fields: Map<String, String>): PackageOptions =
    PackageOptions(
        isFragile = fields["fragile"]?.toBoolean() ?: false,
        insuredValue = fields["insuredValue"]?.let { BigDecimal(it) },
        handling = fields["handling"]
    )
